geopandas/geodataframe.py

```diff
@@ -1588,7 +1597,19 @@ def _constructor(self):
     def _constructor_from_mgr(self, mgr, axes):
         # analogous logic to _geodataframe_constructor_with_fallback
         if not any(isinstance(block.dtype, GeometryDtype) for block in mgr.blocks):
-            return pd.DataFrame._from_mgr(mgr, axes)
+            df = pd.DataFrame._from_mgr(mgr, axes)
+            # operations like loc to append an object row will cast GeometryDtype
+            # blocks to object dtype. Try and cast back
+            # TODO relying on self here is not ideal pandas#56681
+            for c in self.columns[self.dtypes == "geometry"].intersection(df.columns):
+                df[c] = _attempt_ensure_geometry(df[c], crs=self[c].crs)
+            if self._geometry_column_name in df.columns:
+                try:
+                    df = df.set_geometry(self._geometry_column_name, crs=self.crs)
+                except TypeError:
+                    pass
```

Comment:
As a shorter term fix (to just fix the regression for the common case of a "geometry" named columns, and not the longer standing bug), could we just check if the result has a column called "geometry" and in that case try to convert it?

As a way to have something smaller to backport for 0.14.x